src/rt/aaA.d

```diff
@@ -114,442 +266,412 @@ TypeInfo_Struct tiEntry(const TypeInfo keyti, const TypeInfo valti)
+    enum m = 0x5bd1e995;
+    h ^= h >> 13;
+    h *= m;
+    h ^= h >> 15;
+    return h;
+}
 
-    @property const(TypeInfo) keyti() const @safe pure nothrow @nogc
-    { return _keyti; }
+private size_t calcHash(in void* pkey, in TypeInfo keyti)
+{
+    immutable hash = keyti.getHash(pkey);
+    // highest bit is set to distinguish empty/deleted from filled buckets
+    return mix(hash) | HASH_FILLED_MARK;
+}
 
-    // helper function to determine first used bucket, and update implementation's cache for it
-    // NOTE: will not work with immutable AA in ROM, but that doesn't exist yet.
-    size_t firstUsedBucketCache() @safe pure nothrow @nogc
-    in
-    {
-        assert(firstUsedBucket <= buckets.length);
-        foreach(i; 0 .. firstUsedBucket)
-            assert(buckets[i] is null);
-    }
-    body
-    {
-        size_t i;
-        for(i = firstUsedBucket; i < buckets.length; ++i)
-            if(buckets[i] !is null)
-                break;
-        return firstUsedBucket = i;
-    }
+private size_t nextpow2(size_t val) pure nothrow @nogc
+{
+    size_t res = 1;
+    while (res < val)
+        res <<= 1;
+    return res;
 }
 
-/* This is the type actually seen by the programmer, although
- * it is completely opaque.
- */
-struct AA
+private T min(T)(T a, T b) pure nothrow @nogc
 {
-    Impl* impl;
+    return a < b ? a : b;
 }
 
-/**********************************
- * Align to next pointer boundary, so that
- * GC won't be faced with misaligned pointers
- * in value.
- */
-size_t aligntsize(in size_t tsize) @safe pure nothrow @nogc
+private T max(T)(T a, T b) pure nothrow @nogc
 {
-    version (D_LP64) {
-        // align to 16 bytes on 64-bit
-        return (tsize + 15) & ~(15);
-    }
-    else {
-        return (tsize + size_t.sizeof - 1) & ~(size_t.sizeof - 1);
-    }
+    return b < a ? a : b;
 }
 
-extern (C):
+//==============================================================================
+// API Implementation
+//------------------------------------------------------------------------------
 
-/****************************************************
- * Determine number of entries in associative array.
- */
-size_t _aaLen(in AA aa) pure nothrow @nogc
-in
+/// Determine number of entries in associative array.
+extern (C) size_t _aaLen(in AA aa) pure nothrow @nogc
 {
-    //printf("_aaLen()+\n");
-    //_aaInv(aa);
+    return aa ? aa.length : 0;
 }
-out (result)
+
+/// Get LValue for key
+extern (C) void* _aaGetY(AA* aa, const TypeInfo_AssociativeArray ti, in size_t valsz,
+    in void* pkey)
 {
-    size_t len = 0;
+    // lazily alloc implementation
+    if (aa.impl is null)
+        aa.impl = new Impl(ti);
+
+    // get hash and bucket for key
+    immutable hash = calcHash(pkey, ti.key);
 
-    if (aa.impl)
+    // found a value => return it
+    if (auto p = aa.findSlotLookup(hash, pkey, ti.key))
+        return p.entry + aa.keysz + aa.keypad;
+
+    auto p = aa.findSlotInsert(hash);
```

Comment:
This does a lookup very similar to `findSlotLookup` just before. These can be combined, but when trying that, there was only a small improvement.